mlflow/utils/environment.py

```diff
@@ -154,3 +204,74 @@ def _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
             "Only one of `conda_env`, `pip_requirements`, and "
             "`extra_pip_requirements` can be specified"
         )
+
+
+def _is_mlflow_requirement(requirement_string):
+    """
+    Returns True if `requirement_string` represents a requirement for mlflow (e.g. 'mlflow==1.2.3').
+    """
+    return Requirement(requirement_string).name.lower() == "mlflow"
+
+
+def _contains_mlflow_requirement(requirements):
+    """
+    Returns True if `requirements` contains a requirement for mlflow (e.g. 'mlflow==1.2.3').
+    """
+    return any(map(_is_mlflow_requirement, requirements))
+
+
+def _process_pip_requirements(
+    default_pip_requirements, pip_requirements=None, extra_pip_requirements=None
+):
+    """
+    Processes `pip_requirements` and `extra_pip_requirements` passed to `mlflow.*.save_model` or
+    `mlflow.*.log_model`, and returns a tuple of (conda_env, pip_requirements, pip_constraints).
+    """
+    constraints = []
+    if pip_requirements is not None:
+        pip_reqs, constraints = _parse_pip_requirements(pip_requirements)
+    elif extra_pip_requirements is not None:
+        extra_pip_requirements, constraints = _parse_pip_requirements(extra_pip_requirements)
+        pip_reqs = default_pip_requirements + extra_pip_requirements
+    else:
+        pip_reqs = default_pip_requirements
+
+    if not _contains_mlflow_requirement(pip_reqs):
+        pip_reqs.insert(0, "mlflow")
+
+    if constraints:
+        pip_reqs.append(f"-c {_CONSTRAINTS_FILE_NAME}")
+
+    # Set `install_mlflow` to False because `pip_reqs` already contains `mlflow`
+    conda_env = _mlflow_conda_env(additional_pip_deps=pip_reqs, install_mlflow=False)
+    return conda_env, pip_reqs, constraints
+
+
+def _process_conda_env(conda_env):
+    """
+    Processes `conda_env` passed to `mlflow.*.save_model` or `mlflow.*.log_model`, and returns
+    a tuple of (conda_env, pip_requirements, pip_constraints).
+    """
+    if isinstance(conda_env, str):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+    elif isinstance(conda_env, dict):
+        pass
+    else:
+        raise TypeError(
+            "Expected a string path to a conda env yaml file or a `dict` representing a conda env, "
+            "but got `{}`".format(type(conda_env).__name__)
+        )
```

Comment:
Nit: can we replace this with `elif not isinstance(...)`?

Reply:
You're right. Why didn't I come up with this?

Reply:
Done!